File: engines/php/compare.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/carve-src.php';
require_once __DIR__ . '/vendor/autoload.php';
$loader = require getenv('CARVE_PHP_AUTOLOAD') ?: __DIR__ . '/vendor/autoload.php';
carve_bench_apply_src($loader);

use Djot\DjotConverter;
use League\CommonMark\GithubFlavoredMarkdownConverter;
use MarkupCarve\Carve\CarveConverter;

[$script, $engine, $path, $iterationsArg, $trialsArg] = array_pad($argv, 5, null);
$iterations = (int)($iterationsArg ?: 100);
$trials = (int)($trialsArg ?: 5);
$source = file_get_contents($path);
$render = match ($engine) {
    'carve-php' => (function () {
        $converter = new CarveConverter();
        return fn () => $converter->convert($GLOBALS['source']);
    })(),
    'djot-php' => (function () {
        $converter = new DjotConverter();
        return fn () => $converter->convert($GLOBALS['source']);
    })(),
    'league/commonmark-gfm' => (function () {
        $converter = new GithubFlavoredMarkdownConverter();
        return fn () => $converter->convert($GLOBALS['source'])->getContent();
    })(),
    default => throw new RuntimeException("unknown engine: {$engine}"),
};
for ($i = 0; $i < 20; $i++) { $render(); }
$samples = [];
for ($trial = 0; $trial < $trials; $trial++) {
    $start = hrtime(true);
    for ($i = 0; $i < $iterations; $i++) { $render(); }
    $samples[] = (hrtime(true) - $start) / 1e6 / $iterations;
}
$min = min($samples);
$bytes = strlen($source);
$result = [
    'engine' => $engine,
    'bytes' => $bytes,
    'iterations' => $iterations,
    'trials' => $trials,
    'samples' => $samples,
    'ms_per_op' => $min,
    'mb_per_s' => $bytes / 1048576 / ($min / 1000),
    'jit' => (@opcache_get_status(false)['jit']['enabled'] ?? false) === true,
];
// Only the Carve row carries this; on a peer row it would name an engine that
// did not produce the number.
if ($engine === 'carve-php') {
    $result['carve_source'] = carve_bench_describe_src(CarveConverter::class);
}
echo json_encode($result) . "\n";

File: engines/rs/src/compare.rs
use std::fmt::Write as _;
use std::hint::black_box;
use std::time::Instant;

fn main() {
    let mut args = std::env::args().skip(1);
    let engine = args.next().expect("engine");
    let path = args.next().expect("document");
    let iterations: usize = args.next().and_then(|s| s.parse().ok()).unwrap_or(100);
    let trials: usize = args.next().and_then(|s| s.parse().ok()).unwrap_or(5);
    let source = std::fs::read_to_string(path).expect("cannot read document");
    let render = || -> String {
        match engine.as_str() {
            "carve-rs" => carve::to_html(&source),
            "jotdown" => jotdown::html::render_to_string(jotdown::Parser::new(&source)),
            "comrak" => {
                let mut options = comrak::Options::default();
                options.extension.table = true;
                comrak::markdown_to_html(&source, &options)
            }
            "pulldown-cmark" => {
                let mut html = String::new();
                let options = pulldown_cmark::Options::ENABLE_TABLES;
                pulldown_cmark::html::push_html(
                    &mut html,
                    pulldown_cmark::Parser::new_ext(&source, options),
                );
                html
            }
            _ => panic!("unknown engine: {engine}"),
        }
    };
    for _ in 0..20 {
        black_box(render());
    }
    let mut samples = Vec::with_capacity(trials);
    for _ in 0..trials {
        let start = Instant::now();
        for _ in 0..iterations {
            black_box(render());
        }
        samples.push(start.elapsed().as_secs_f64() * 1000.0 / iterations as f64);
    }
    let min = samples.iter().copied().fold(f64::INFINITY, f64::min);
    let throughput = source.len() as f64 / 1_048_576.0 / (min / 1000.0);
    let mut encoded = String::new();
    for (index, sample) in samples.iter().enumerate() {
        if index > 0 {
            encoded.push(',');
        }
        write!(&mut encoded, "{sample:.6}").unwrap();
    }
    // Only the Carve row carries `carve_source`; on a peer row it would name an
    // engine that did not produce the number.
    let carve_source = if engine == "carve-rs" {
        format!(",\"carve_source\":\"{}\"", env!("CARVE_ENGINE_SOURCE"))
    } else {
        String::new()
    };
    println!("{{\"engine\":\"{engine}\",\"bytes\":{},\"iterations\":{iterations},\"trials\":{trials},\"samples\":[{encoded}],\"ms_per_op\":{min:.6},\"mb_per_s\":{throughput:.2}{carve_source}}}", source.len());
}
